enhancements/ingress/client-tls.md

```diff
@@ -0,0 +1,325 @@
+---
+title: client-tls
+authors:
+  - "@Miciah"
+reviewers:
+  - "@danehans"
+  - "@frobware"
+  - "@knobunc"
+  - "@sgreene570"
+approvers:
+  - "@knobunc"
+creation-date: 2020-08-31
+last-updated: 2020-08-31
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Ingress Client TLS API
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement extends the IngressController API to allow the user to
+configure the use of client certificates for mutual TLS.
+
+## Motivation
+
+Some applications require the use of TLS with client certificates in addition to
+server certificates in order to provide two-way certificate-based authentication
+("mutual TLS", or "mTLS").  OpenShift 3 supports mTLS in the router, and the
+absence of mTLS support in OpenShift 4 blocks some users from upgrading.
+
+### Goals
+
+1. Enable the user to configure a client CA certificate bundle.
+2. Enable the user to configure optional or mandatory use of client certificates.
+3. Respect client CA certificates' certificate revocation lists.
+4. Allow the cluster administrator to restrict certificates by subject name.
```

Comment:
What about Online Certificate Status Protocol (OCSP) for checking certificate revocation? 
What about by validating Issuer CN? 

In short the logic behind verifying a 'certificate' can be different between customers; and if we don't provide a system is flexible we run the risk of just backlogging RFE's for each customer situation. 

HTTPD uses a lexicon (https://httpd.apache.org/docs/2.4/mod/mod_ssl.html#sslrequire) to allow for a wide degree of flexibility in verifying the certificate. It doesn't seem as though haproxy provides the same level of verification! http://cbonte.github.io/haproxy-dconv/2.0/configuration.html#5.1-verify and can only verify clients based on CA valitity,a s configured in the endpoitns 'ca-file' or 'crl-file'directive.

Reply:
I don't believe HAProxy has the capability to perform OCSP lookups.  OCSP is usually implemented in clients; I don't know that proxies typically implement OCSP (looks like Envoy doesn't implement it: https://www.github.com/envoyproxy/envoy/issues/6009).

Matching against the issuer DN (including the CN) would be feasible, though it would be a step beyond what we supported in OCP 3.  I cannot commit to including it in this enhancement, but I won't rule it out at this time.

Reply:
nginx has always supported CRL but only very recently added OCSP in 1.19.0 - generally proxies don't want to implement it as it'll be blocking operation and make the website seem slow to load.

I think it's acceptable to leave all but the most trivial DN validation to the customer, they'll have the DN etc passed in headers so could implement custom logic. However they wouldn't be able to perform revocation checks, hence why that needs to be implemented for them.

Couldn't you restrict Issuer CN by only adding the the issuers you want to the certificate bundle (ie. the ability to supply an intermediate CA rather than just root CA's)?

Reply:
> nginx has always supported CRL but only very recently added OCSP in 1.19.0 - generally proxies don't want to implement it as it'll be blocking operation and make the website seem slow to load.

Right, OCSP doesn't seem suitable for proxies.  

> I think it's acceptable to leave all but the most trivial DN validation to the customer, they'll have the DN etc passed in headers so could implement custom logic. However they wouldn't be able to perform revocation checks, hence why that needs to be implemented for them.

Right, this enhancement doesn't describe it, but the HAProxy configuration template adds some HTTP headers that backend applications can use to validate the client:
```
  # Add X-SSL* headers to pass client certificate information to the backend. http-request set-header X-SSL                  %[ssl_fc]
  http-request set-header X-SSL-Client-Verify    %[ssl_c_verify]
  http-request set-header X-SSL-Client-Serial    %{+Q}[ssl_c_serial,hex]
  http-request set-header X-SSL-Client-Version   %{+Q}[ssl_c_version]
  http-request set-header X-SSL-Client-SHA1      %{+Q}[ssl_c_sha1,hex]
  http-request set-header X-SSL-Client-DN        %{+Q}[ssl_c_s_dn]
  http-request set-header X-SSL-Client-CN        %{+Q}[ssl_c_s_dn(cn)]
  http-request set-header X-SSL-Issuer           %{+Q}[ssl_c_i_dn]
  http-request set-header X-SSL-Client-NotBefore %{+Q}[ssl_c_notbefore]
  http-request set-header X-SSL-Client-NotAfter  %{+Q}[ssl_c_notafter]
  http-request set-header X-SSL-Client-DER       %{+Q}[ssl_c_der,base64]
```

> Couldn't you restrict Issuer CN by only adding the the issuers you want to the certificate bundle (ie. the ability to supply an intermediate CA rather than just root CA's)?

We configure HAProxy using `verify ca-file`, and based on HAProxy's documentation, it looks like the provided certificates can be root or intermediate CAs, but I would need to do further testing to be certain.